api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java

```diff
@@ -0,0 +1,151 @@
+ * "aud": "client id",
+ * "c_hash": "",
+ * "acr": "https://stsstg.nih.gov/assurance/ial/1 https://stsstg.nih.gov/assurance/aal/1",
+ * "azp": "",
+ * "auth_time": 1615910504,
+ * "iss": "https://stsstg.nih.gov",
+ * "exp": 1615996939,
+ * "iat": 1615910539
+ * }</pre>
+ *
+ * The acr claim contains user IAL status, the number after /assurance/ial/
+ *
+ * <p>Step3: User access_token to pull RAS user info. A sample userInfo in Json format:
+ *
+ * <pre>{@code
+ * "sub":"subId",
+ * "name" : "",
+ * "preferred_username":"user1@Login.Gov",
+ * "userId":"user1"
+ * "email":"foo@gmail.com"
+ * }</pre>
+ *
+ * The {@code preferred_username} field should end with "@login.gov" if using that to login. The
+ * {@code email} field is the user used to sign in in login.gov. We can use that as login.gov user
+ * name.
+ *
+ * <p>Step4: Use step3's login.gov username to update AoU database by {@link
+ * UserService#updateRasLinkLoginGovStatus(String)}. Then return it as user profile.
+ *
+ * <p>TODO(yonghao): Fow now we return {@llink ForbiddenException} for all scenarios, determine if
+ * we need to differentiate IAL vs Login.gov scenarios, and give that information to UI.
+ */
+@Service
+public class RasLinkService {
+  private static final Logger log = Logger.getLogger(RasLinkService.class.getName());
+
+  private final UserService userService;
+  private final Provider<OpenIdConnectClient> rasOidcClientProvider;
+
+  @Autowired
+  public RasLinkService(
+      UserService userService,
+      @Qualifier(RAS_OIDC_CLIENT) Provider<OpenIdConnectClient> rasOidcClientProvider) {
+    this.userService = userService;
+    this.rasOidcClientProvider = rasOidcClientProvider;
+  }
+
+  /** Links RAS login.gov account with AoU account. */
+  public DbUser linkRasLoginGovAccount(String authCode, String redirectUrl) {
+    OpenIdConnectClient rasOidcClient = rasOidcClientProvider.get();
+    JsonNode userInfoResponse;
+    try {
+      // Oauth dance to get id token and access token.
+      TokenResponse tokenResponse =
+          rasOidcClient.codeExchange(authCode, redirectUrl, RAS_AUTH_CODE_SCOPES);
+
+      // Validate IAL status.
+      String acrClaim =
+          decodedJwt(tokenResponse.get(Id_TOKEN_FIELD_NAME).toString())
+              .getClaim(ACR_CLAIM)
+              .asString();
+      if (!isIal2(acrClaim)) {
+        throw new ForbiddenException(
+            String.format("User does not have IAL2 enabled, acrClaim: %s", acrClaim));
+      }
+      // Fetch user info.
+      userInfoResponse = rasOidcClient.fetchUserInfo(tokenResponse.getAccessToken());
+    } catch (IOException e) {
+      log.log(Level.WARNING, "Failed to link RAS account", e);
+      throw new ServerErrorException("Failed to link RAS account", e);
+    }
+    return userService.updateRasLinkLoginGovStatus(getLoginGovUsername(userInfoResponse));
+  }
+
+  /** Validates user has IAL2 setup. See class javadoc Step2 for more details. */
+  static boolean isIal2(String acrClaim) {
```

Comment:
You said this is string delimited - then would it make sense to first break this string into individual tokens, and verify that at least one matches exactly?

It's unlikely to matter since presumably these ACR values are maintained by RAS, but it seems incorrect that the following claims are accepted by this regex check:
"myevilsite.com/https://stsstg.nih.gov/assurance/ial/2"
"https://stsstg.nih.gov/assurance/ial/27"

Reply:
Changed to the exactly match. Thanks